src/Symfony/Component/HttpFoundation/InputBag.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+use Symfony\Component\HttpFoundation\Exception\BadRequestException;
+
+/**
+ * InputBag is a container for user input values such as $_GET, $_POST, $_REQUEST, and $_COOKIE.
+ *
+ * @author Saif Eddin Gmati <saif.gmati@symfony.com>
+ */
+final class InputBag extends ParameterBag
+{
+    /**
+     * Returns a string input value by name.
+     *
+     * @param string|null $default The default value if the input key does not exist
+     *
+     * @return string|null
+     */
+    public function get(string $key, $default = null)
+    {
+        if (null !== $default && !is_scalar($default) && !\is_callable([$default, '__toString'])) {
+            @trigger_error(sprintf('Passing a non-string value as 2nd argument to "%s::%s()" is deprecated since Symfony 5.1, pass a string or null instead.', __CLASS__, __METHOD__), E_USER_DEPRECATED);
+        }
+
+        $value = parent::get($key, $this);
+
+        if (null !== $value && $this !== $value && !is_scalar($value) && !\is_callable([$value, '__toString'])) {
+            @trigger_error(sprintf('Retrieving a non-string value from "%s::%s()" is deprecated since Symfony 5.1, and will throw a "%s" exception in Symfony 6.0, use "%1$s::all()" instead.', __CLASS__, __METHOD__, BadRequestException::class), E_USER_DEPRECATED);
+        }
+
+        return $this === $value ? $default : $value;
+    }
+
+    /**
+     * Returns the inputs.
+     *
+     * @param string|null $key The name of the input to return or null to get them all
+     */
+    public function all(string $key = null): array
```

Comment:
please use a different method name for retrieving all parameters and retrieving a single param that is meant to be an array. those are different concepts that should not simply depend on the parameter

Reply:
The proposed signature is consistent with `ResponseHeaderBag::all(string $key = null)`.

Reply:
(and with `Symfony\Component\Messenger\Envelope::all(string $stampFqcn = null)`, even if that's in another component)

Reply:
I know. I'm repeating myself: https://github.com/symfony/symfony/pull/32122#discussion_r312195599
This PR tries to make types safer. But by depending on arguments to execute logical different things and even return different types, it does not make it better.

Reply:
People need to know that depending on the argument, calling the method has totally different outcomes. That is not self-exlpaining code. And it's not statically analazable which is what was meant to be fixed in #31317

Reply:
Everywhere i have to put code like

```
/** @var StampInterface[] $stamps */
        foreach ($envelope->all() as $stamps) {
```

because otherwise phpstorm, phpstan etc will complain.

Reply:
I don't totally agree but I understand. We need to keep consistency for now. If you want to change these methods, it should be a separate PR with a BC layer. Having responses have a different style than requests is no go of course.

Reply:
@Tobion i agree that other getters ( e.g `getInt` ) should be fixed to throw a bad request exception too. however, this won't be done in this PR ( and i am not sure if we should deprecate getting non-`x` values from `getX` first )

Reply:
The signature here matches `ResponseHeaderBag::all`, `ParameterBag::all` and other `all` methods. disallowing retrieving arrays here without introducing `getArray` would be an issue. i believe if we are going to do such change, it should be done everywhere, and carefully. this is outside the scope of this PR.